inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/SPARQLQueryStore.java

```diff
@@ -69,34 +72,87 @@ private static final String optionalLanguageFilteredValue(String aProperty, Stri
+            String aLanguage, String variable, String filterVariable)
+    {
+        StringBuilder fragment = new StringBuilder();
+        for (KBHandle label : labelProperties) {
+            fragment.append(optionalLanguageFilteredValue("<" + label.getIdentifier() + ">",
+                    aLanguage, variable, filterVariable));
+            fragment.append(optionalLanguageFilteredValue("<" + label.getIdentifier() + ">", null,
+                    variable, filterVariable));
+        }
+        return fragment.toString();
+    }
+    
     
+    private static final String languageFilterForVariable(String aVariable, String aLanguage) {
+        StringBuilder fragment = new StringBuilder();
+        if (aLanguage != null) {
+            // If a certain language is specified, we look exactly for that
+            String escapedLang = NTriplesUtil.escapeString(aLanguage);
+            fragment.append("    FILTER(LANGMATCHES(LANG(").append(aVariable).append("), \"").append(escapedLang).append("\"))\n");
+        }
+        else {
+            // If no language is specified, we look for statements without a language as otherwise
+            // we might easily run into trouble on multi-lingual resources where we'd get all the
+            // labels in all the languages being retrieved if we simply didn't apply any filter.
+            fragment.append("    FILTER(LANG(").append(aVariable).append(") = \"\")\n");
+        }
+        return fragment.toString();
+    }
+
     /** 
      * Query to list all concepts from a knowledge base.
      */
-    public static final String queryForAllConceptList(KnowledgeBase aKB)
+    public static final String queryForAllConceptList(KnowledgeBase aKB,
+            Set<KBHandle> labelProperties)
     {
         return String.join("\n"
                 , SPARQL_PREFIX
-                , "SELECT DISTINCT ?s ?l ?labelGeneral WHERE { "
+                , "SELECT DISTINCT ?s ?l ?labelGeneral ?spl WHERE { "
                 , "  { ?s ?pTYPE ?oCLASS . } "
                 , "  UNION { ?someSubClass ?pSUBCLASS ?s . } ."
                 , optionalLanguageFilteredValue("?pLABEL", aKB.getDefaultLanguage(),"?s","?l")
                 , optionalLanguageFilteredValue("?pLABEL", null,"?s","?labelGeneral")
+                , optionalSubPropertyLabelFilter(labelProperties, aKB.getDefaultLanguage(),"?s","?spl")
                 , "}"
                 , "LIMIT " + aKB.getMaxResults());
     }
     
     /** 
      * Query to list all instances from a knowledge base.
      */
-    public static final String listInstances(KnowledgeBase aKB)
+    public static final String listInstances(KnowledgeBase aKB, Set<KBHandle> labelProperties)
     {
         return String.join("\n"
                 , SPARQL_PREFIX
-                , "SELECT DISTINCT ?s ?l ?labelGeneral WHERE {"
+                , "SELECT DISTINCT ?s ?l ?labelGeneral ?spl WHERE {"
```

Comment:
We have `?l` (label with language) and `?labelGeneral` (label without language). Why do we need an additional `?spl`? IMHO the `queryForOptionalLabel` should be overlapping with `optionalLanguageFilteredValue`, so let's bind `queryForOptionalLabel ` to `?l` and drop `?spl`.

Reply:
then the problem would be that it can at times ignore the label present with rdfs:label and take the label provided with its sub property label. Thats why I did it separate (for the worst case scenario)

Reply:
Why would it ignore the rdfs:label?

Reply:
Because if we take the same variable, that would be a possibility. For now, in the list of sub property label list we do not include label.